src/Npgsql/ConnectorPool.cs

```diff
@@ -472,23 +484,71 @@ void ReleaseOneWaiter()
                     break;
         }
 
+        // Manual reactivation of timer happens in callback
+        void EnablePruning()
+        {
+            lock (_pruningTimer)
+            {
+                _pruningTimerEnabled = true;
+                _pruningTimer.Change(_pruningSamplingInterval, Timeout.InfiniteTimeSpan);
+            }
+        }
+
+        void DisablePruning()
+        {
+            lock (_pruningTimer)
+            {
+                _pruningTimer.Change(Timeout.Infinite, Timeout.Infinite);
+                _pruningSampleIndex = 0;
+                _pruningTimerEnabled = false;
+            }
+        }
+
         static void PruneIdleConnectors(object? state)
         {
             var pool = (ConnectorPool)state!;
-            var idle = pool._idle;
-            var now = DateTime.UtcNow;
-            var idleLifetime = pool.Settings.ConnectionIdleLifetime;
+            var samples = pool._pruningSamples;
+            int toPrune;
+            lock (pool._pruningTimer)
+            {
+                // Check if we might have been contending with DisablePruning.
+                if (!pool._pruningTimerEnabled)
+                    return;
+
+                var sampleIndex = pool._pruningSampleIndex;
+                if (sampleIndex < pool._pruningSampleSize)
+                {
+                    samples[sampleIndex] = pool.State.Idle;
+                    pool._pruningSampleIndex = sampleIndex + 1;
+                    pool._pruningTimer.Change(pool._pruningSamplingInterval, Timeout.InfiniteTimeSpan);
+                    return;
+                }
+
+                // Reset before releasing the lock and starting the prune
+                toPrune = Median(samples);
+                pool._pruningSampleIndex = 0;
+                pool._pruningTimer.Change(pool._pruningSamplingInterval, Timeout.InfiniteTimeSpan);
+            }
 
+            var idle = pool._idle;
             for (var i = 0; i < idle.Length; i++)
             {
-                if (Volatile.Read(ref pool.State.Open) <= pool._min)
+                if (Volatile.Read(ref pool.State.Open) <= pool._min || toPrune == 0)
                     return;
 
                 var connector = idle[i];
-                if (connector == null || (now - connector.ReleaseTimestamp).TotalSeconds < idleLifetime)
+                if (connector == null || Interlocked.CompareExchange(ref idle[i], null, connector) != connector)
```

Comment:
Can't this be simplified by just doing Interlocked.Exchange, setting `idle[i]` to null and checking what came out of it? If it was null we failed to remove a connector, if it was non-null we got one and can close.

Reply:
Good point, not sure why this is here in this way